{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b448e488",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# Testing params\n",
    "excel_path=r\"C:\\Users\\lwoods\\Documents\\LW_Projects_folder\\01_MO\\G02_Barbacid\\R12_Oksana\\P01_KRS1\\E03_1gel_2_isoforms\\CNIO_prot_core_results\\MaxQuant_protein_groups_results.xlsx\"\n",
    "fasta_file=r\"C:\\Users\\lwoods\\Documents\\LW_Projects_folder\\01_MO\\G02_Barbacid\\R12_Oksana\\P01_KRS1\\E03_1gel_2_isoforms\\fasta_folder\\merged_fasta_folder\\custom_fasta_merged_w_HUMAN_database.fasta\"\n",
    "POIs=['P23246', 'Q9BUJ2', 'P53992', 'Q00839', 'Q8IVT5', 'A0A2R8Y4X0', 'A0A2R8Y5H9']\n",
    "experiments=['Down','Up']\n",
    "identification=\"Razor + Unique\"\n",
    "colours=['#FF5900', '#DA16FF', '#1CA71C', '#2E91E5', '#E15F99', '#25B0B0', '#B68100', '#750D86', '#EB663B', '#511CFB', '#00A08B', '#FB00D1', '#FC0080', '#B2828D', '#6C7C32', '#778AAE', '#862A16', '#A777F1', '#620042', '#1616A7', '#DA60CA', '#6C4516', '#0D2A63', '#AF0038']\n",
    "reattribute_peptides=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53714b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from nbconvert import HTMLExporter\n",
    "from nbformat import read, write\n",
    "from pathlib import Path\n",
    "import os\n",
    "from IPython.display import Markdown, display\n",
    "import pandas as pd\n",
    "import plotly\n",
    "import plotly.express as px\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import plotly.graph_objects as go\n",
    "import re\n",
    "from matplotlib.colors import to_rgb\n",
    "\n",
    "# Report text stuff. Looks a bit dumb right now\n",
    "# so TODO: adjust text parameters, etc. to make\n",
    "# a nicer report\n",
    "# Also, all figures should be output as standalone htmls\n",
    "# so need to include output folder as parameter\n",
    "\n",
    "def printmd(string):\n",
    "    from IPython.display import Markdown, display\n",
    "    markdown_string=display(Markdown(string))\n",
    "\n",
    "\n",
    "def format_tick(n: float):\n",
    "    m, e = f\"{n:.0e}\".split(\"e\")\n",
    "    e = int(e)\n",
    "    return f\"{m} x 10<sup>{e}</sup>\"\n",
    "\n",
    "\n",
    "def extract_protein_names(row, prot_id_col, POIs):\n",
    "    headers = row['Fasta headers'].split(';') if not pd.isna(row['Fasta headers']) else []\n",
    "    majority_ids = row[prot_id_col].split(';')\n",
    "\n",
    "    names = []\n",
    "    for header, majority_id in zip(headers, majority_ids):\n",
    "        if majority_id in POIs:\n",
    "            names.append(majority_id)\n",
    "        else:\n",
    "            parts = header.split('|')\n",
    "            if len(parts) >= 3:\n",
    "                try:\n",
    "                    identifier, description = parts[2].split(\" \", 1)\n",
    "                    names.append(identifier)\n",
    "                except:\n",
    "                    names.append(majority_id)\n",
    "            else:\n",
    "                names.append(majority_id)\n",
    "\n",
    "    names.extend(majority_ids[len(headers):])\n",
    "\n",
    "    return ';'.join(names)\n",
    "\n",
    "# Get df of POIs but make new col for names for plotting\n",
    "# (possibly more trouble than it's worth....)\n",
    "def get_POIs_df(experiment, protein_groups, num, POIs):\n",
    "    protein_groups['Plotting names'] = protein_groups.apply(extract_protein_names, args=(prot_id_col, POIs), axis=1)\n",
    "    mask = protein_groups['Plotting names'].str.split(';').apply(lambda x: any(poi in x for poi in POIs))\n",
    "    POIs_df = protein_groups[mask]\n",
    "    return POIs_df\n",
    "\n",
    "# Get the top n proteins (after subtracting POIs)\n",
    "# n is buried right now, should be more accessible\n",
    "# and also may be better to use intensity thresh instead\n",
    "def protein_df_organization(experiment,protein_groups,num,POIs_df,prot_id_col):\n",
    "    num_POIs=len(POIs_df)\n",
    "    intensity_col='Intensity '+str(experiment)\n",
    "    head_n=num-num_POIs\n",
    "    protein_groups['Plotting names'] = protein_groups.apply(extract_protein_names, args=(prot_id_col,POIs), axis=1)\n",
    "    total_intensity=protein_groups[intensity_col].sum()\n",
    "    top_n_intensity=protein_groups[~protein_groups[prot_id_col].isin(POIs_df[prot_id_col])].sort_values(intensity_col, ascending=False).head(head_n)\n",
    "    top_n=pd.concat([POIs_df, top_n_intensity], axis=0)\n",
    "    top_n['Relative protein intensity '+str(experiment)+' [%]']=top_n[intensity_col]/total_intensity*100\n",
    "    return top_n\n",
    "\n",
    "def make_report_title(experiment,POIs,identification):\n",
    "    POI_string1=(', ').join(POIs[0:-2])\n",
    "    POI_string2=(' and ').join(POIs[-2:])\n",
    "    if len(POIs)>2:\n",
    "        POI_string_full=(', ').join([POI_string1,POI_string2])\n",
    "    else:\n",
    "         POI_string_full=POI_string2\n",
    "    printmd(f'<h1 style=\"font-family:Open Sans, verdana, arial, sans-serif;font-weight:normal;\">Report for MS analysis of {experiment}: {POI_string_full}</h1>')\n",
    "    if identification==\"Unique\":\n",
    "        printmd(f'<span style=\"font-family:Open Sans, verdana, arial, sans-serif;font-size:12;margin-top:0px;margin-bottom:0px;\">Note that this report only considers peptides unique to the protein of interest.<br><br><br></span>')\n",
    "\n",
    "def make_subtitle(num,POIs,POIs_df):\n",
    "    num_POIs=len(POIs_df)\n",
    "    new_POIs=POIs_df['Plotting names'].values.tolist()\n",
    "    POI_string1=(', ').join(new_POIs[0:-2])\n",
    "    POI_string2=(' and ').join(new_POIs[-2:])\n",
    "    if len(new_POIs)>2:\n",
    "        POI_string_full=(', ').join([POI_string1,POI_string2])\n",
    "    else:\n",
    "        POI_string_full=POI_string2\n",
    "    unidentified_POIs=[]\n",
    "    for POI in POIs:\n",
    "        is_present = POIs_df['Plotting names'].apply(lambda x: POI in x.split(';')).any()\n",
    "        if is_present == False:\n",
    "            unidentified_POIs.append(POI)\n",
    "    if len(unidentified_POIs)<len(POIs):\n",
    "        printmd(f'<h2 style=\"font-family:Open Sans, verdana, arial, sans-serif;font-weight:normal;margin-top:0px;margin-bottom:0px;\">Protein signal intensities</h2><span style=\"font-family:Open Sans, verdana, arial, sans-serif;font-size:12;margin-top:0px;margin-bottom:0px;\">The following displays the raw signal intensities corresponding to {POI_string_full} with the top {num-num_POIs} proteins.<span>')\n",
    "    else:\n",
    "        printmd(f'<h2 style=\"font-family:Open Sans, verdana, arial, sans-serif;font-weight:normal;margin-top:0px;margin-bottom:0px;\">Protein signal intensities</h2><span style=\"font-family:Open Sans, verdana, arial, sans-serif;font-size:12;margin-top:0px;margin-bottom:0px;\">The following displays the raw signal intensities corresponding to the top {num-num_POIs} proteins.<span>')\n",
    "    if len(unidentified_POIs)>0:\n",
    "        unid_POIs_string=', '.join(unidentified_POIs)\n",
    "        printmd(f'<span style=\"font-family:Open Sans, verdana, arial, sans-serif;font-size:12;margin-top:0px;margin-bottom:0px;\">The following proteins of interest were not identified in any experiment: {unid_POIs_string}.<span>')\n",
    "\n",
    "def shorten_name(name, max_len=26):\n",
    "    return name[:max_len-3] + '...' if len(name) > max_len else name        \n",
    "\n",
    "def topn_proteins_identified(top_n,POIs,POIs_df,num,experiment,colours,prot_id_col):\n",
    "    leg_names=top_n['Plotting names'].values.tolist()\n",
    "    top_n['Plotting names'] = top_n['Plotting names'].apply(lambda x: shorten_name(x))\n",
    "    \n",
    "    group_colour_mapping = {}\n",
    "    for poi, color in zip(sorted(POIs), colours):\n",
    "        group_colour_mapping[poi] = color\n",
    "\n",
    "    for majority_prot_list in top_n['Plotting names']:\n",
    "        proteins = majority_prot_list.split(\";\")\n",
    "        if any(\"CON__\" in protein for protein in proteins):\n",
    "            group_colour_mapping[majority_prot_list] = '#a6a3a2'\n",
    "        elif majority_prot_list not in POIs:\n",
    "            if any(poi in proteins for poi in POIs):\n",
    "                for poi in POIs:\n",
    "                    if poi in proteins:\n",
    "                        group_colour_mapping[majority_prot_list] = group_colour_mapping[poi]\n",
    "                        break\n",
    "            else:\n",
    "                group_colour_mapping[majority_prot_list] = 'black'\n",
    "    \n",
    "    intensity_col='Intensity '+str(experiment)\n",
    "    num_POIs=len(POIs_df)\n",
    "    POIs_plotting=POIs_df['Plotting names'].values.tolist()\n",
    "\n",
    "    def format_text(value):\n",
    "        float_value = float(value)\n",
    "        if float_value == 0.0:\n",
    "            return \"Absent\"\n",
    "        elif round(float_value, 2) == 0:\n",
    "            return \"<0.01%\"\n",
    "        else:\n",
    "            return f'{round(float_value, 2)}%'\n",
    "\n",
    "    text_labels = [format_text(val) for val in top_n['Relative protein intensity '+str(experiment)+' [%]']]\n",
    "    \n",
    "    columns_to_fill = [prot_id_col, 'Gene names', 'Protein names', 'Potential contaminant', 'MS/MS count']     \n",
    "    top_n['Potential contaminant'] = top_n['Potential contaminant'].map({'+': 'Yes', '': 'No'}).fillna('No')\n",
    "\n",
    "    # Handling protein groups\n",
    "    def extract_gene_names(header):\n",
    "        if \";\" in str(header):\n",
    "            headers = str(header).split(\";\")\n",
    "        else:\n",
    "            headers = [str(header)]\n",
    "        gene_names = [re.search(r'GN=([\\w\\d]+)', h).group(1) if re.search(r'GN=([\\w\\d]+)', h) else '' for h in headers]\n",
    "        return \";\".join(gene_names)\n",
    "\n",
    "    def extract_protein_names(header):\n",
    "        if \";\" in str(header):\n",
    "            headers = str(header).split(\";\")\n",
    "        else:\n",
    "            headers = [str(header)]\n",
    "        protein_names = [re.search(r'\\|([^|]+)\\s+OS=', h).group(1) if re.search(r'\\|([^|]+)\\s+OS=', h) else '' for h in headers]\n",
    "        return \";\".join(protein_names)\n",
    "    \n",
    "    if 'Gene names' not in top_n.columns:\n",
    "        top_n['Gene names'] = top_n['Fasta headers'].apply(extract_gene_names)\n",
    "\n",
    "    if 'Protein names' not in top_n.columns:\n",
    "        top_n['Protein names'] = top_n['Fasta headers'].apply(extract_protein_names)\n",
    "    \n",
    "    for col in columns_to_fill:\n",
    "        top_n[col].fillna(\"\", inplace=True)\n",
    "\n",
    "    fig = px.bar(top_n, \n",
    "                 x=top_n['Plotting names'], \n",
    "                 y=top_n[intensity_col], \n",
    "                 color=top_n['Plotting names'], \n",
    "                 color_discrete_map=group_colour_mapping, \n",
    "                 text=text_labels,\n",
    "                 hover_data={'Plotting names': False, intensity_col: True},\n",
    "                 custom_data=[\n",
    "                              prot_id_col, 'Gene names', 'Protein names','Potential contaminant',\n",
    "                              'MS/MS count',\n",
    "                             ])\n",
    "    # Possible TODO: make the hoverables inputs, but may be a bit too much\n",
    "    hovertemplate = (\n",
    "        \"<b>Protein/Protein group</b>: %{x}<br>\"\n",
    "        \"<b>Intensity</b>: %{y}<br>\"\n",
    "        \"<b>Prot ID</b>: %{customdata[0]}<br>\"\n",
    "        \"<b>Gene Names</b>: %{customdata[1]}<br>\"\n",
    "        \"<b>Protein Names</b>: %{customdata[2]}<br>\"\n",
    "        \"<b>Potential contaminant</b>: %{customdata[3]}<br>\"\n",
    "        \"<b>MS/MS count</b>: %{customdata[4]}<br>\"\n",
    "    )\n",
    "\n",
    "    fig.update_traces(texttemplate='%{text}', \n",
    "                      textposition='outside', \n",
    "                      hovertemplate=hovertemplate)\n",
    "    \n",
    "    min_value = top_n[intensity_col].min()\n",
    "    max_value = top_n[intensity_col].max()\n",
    "    y_axis_limit = max_value + 0.10 * max_value\n",
    "\n",
    "    fig.update_layout(\n",
    "        title=f'Signal intensity (peptide sum) of top {num} proteins identified<br><sup>Bars are marked with relative protein intensity. Known contaminants are marked in grey and other proteins in black.<sup>',\n",
    "        yaxis_title=\"Intensity\", \n",
    "        xaxis_title=\"Protein group\",\n",
    "        plot_bgcolor='#EEEAE8',\n",
    "        legend_title_text='Proteins',\n",
    "        yaxis={'tickformat':\"e\", 'range': [0, y_axis_limit], 'ticks': 'outside'},\n",
    "        font=dict(color='black',size=12)\n",
    "    )\n",
    "\n",
    "    config = {\n",
    "      'toImageButtonOptions': {\n",
    "        'format': 'png',\n",
    "        'filename': 'Protein_report_plot',\n",
    "        'scale': 4\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    fig.show(renderer=\"notebook\", config=config)\n",
    "\n",
    "    \n",
    "def filter_peptides(peptides_df, protein_groups_df, quantification,prot_id_col,pept_id_col):\n",
    "    final_peptides_df = pd.DataFrame()\n",
    "\n",
    "    # Iterating (:/)--can't think of a better way round rn\n",
    "    for _, row in protein_groups_df.iterrows():\n",
    "        protein_group = row[prot_id_col]\n",
    "        if quantification == \"All\":\n",
    "            filtered_peptides = peptides_df[peptides_df[pept_id_col].str.contains(protein_group)]\n",
    "        elif quantification == \"Unique\":\n",
    "            filtered_peptides = peptides_df[peptides_df[pept_id_col].str.contains(protein_group) & (peptides_df['Unique (Groups)'] == 'yes')]\n",
    "        elif quantification == \"Razor + Unique\":\n",
    "            # Now need to grab the \"is razor\" list to sort through so I can properly determine\n",
    "            # which peptides to show depending on quant mode\n",
    "            #razor_status = row['Peptide is razor'].split(\";\")\n",
    "            #peptide_ids = row['Peptide IDs'].split(\";\")\n",
    "            peptide_is_razor = [x == 'True' for x in row['Peptide is razor'].split(';')]\n",
    "            peptide_ids = list(map(int, row['Peptide IDs'].split(';')))\n",
    "            current_protein_peptides = peptides_df[peptides_df['id'].isin(peptide_ids)].copy()\n",
    "            current_protein_peptides['is_razor'] = current_protein_peptides['id'].map(dict(zip(peptide_ids, peptide_is_razor)))\n",
    "            filtered_peptides = current_protein_peptides[(current_protein_peptides['is_razor']) | \n",
    "                                (current_protein_peptides['Unique (Groups)'] == 'yes')]\n",
    "            \n",
    "        # Accumulate dfs\n",
    "        final_peptides_df = pd.concat([final_peptides_df, filtered_peptides])\n",
    "\n",
    "    return final_peptides_df\n",
    "\n",
    "    \n",
    "def peptide_plots(fasta_file,peptides_df,POIs,protein_df,exp,colours,prot_id_col,pept_id_col):\n",
    "    def read_fasta(fasta_file):\n",
    "        # Assumes uniprot--should make adjustable\n",
    "        fasta_dict ={}\n",
    "        for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            protein_id = record.id.split(\"|\")[1]  \n",
    "            fasta_dict[protein_id]=[str(record.seq),len(record.seq)]\n",
    "        return fasta_dict\n",
    "\n",
    "    def filter_peptides(peptides_df, exp):\n",
    "        df = peptides_df[peptides_df[f'Intensity {exp}'] != 0]\n",
    "        return df\n",
    "\n",
    "    def get_protein_peptides(peptide_df, protein_df, exp, prot_id_col, pept_id_col):\n",
    "        protein_peptide_dict = {}\n",
    "        # Loop through protein groups\n",
    "        for _, row in protein_df.iterrows():\n",
    "            proteins = row[prot_id_col].split(';')\n",
    "            peptide_ids = row['Peptide IDs'].split(';')\n",
    "            for protein in proteins:\n",
    "                valid_peptides = []\n",
    "                for peptide_id in peptide_ids:\n",
    "                    peptide_row = peptide_df[peptide_df['id'] == int(peptide_id)]\n",
    "                    if peptide_row.empty:\n",
    "                        continue\n",
    "                    if protein in peptide_row[pept_id_col].iloc[0].split(';'):\n",
    "                        peptide_sequence = peptide_row['Sequence'].iloc[0]\n",
    "                        intensity = peptide_row[f'Intensity {exp}'].iloc[0]\n",
    "                        # Find start and end of peptide within protein sequence\n",
    "                        # Remember to correct for Python indexing!\n",
    "                        start = seq_dict[protein][0].find(peptide_sequence) + 1 \n",
    "                        end = start + len(peptide_sequence) - 1\n",
    "                        if start > 0: \n",
    "                            valid_peptides.append((start, end, intensity, peptide_sequence))\n",
    "                protein_peptide_dict[protein] = valid_peptides\n",
    "        return protein_peptide_dict\n",
    "\n",
    "    def combine_ranges(ranges):\n",
    "        # This is for \"gluing\" overlapping and touching peptides\n",
    "        if not ranges: \n",
    "            return []\n",
    "\n",
    "        ranges = sorted(ranges, key=lambda x: x[0])  \n",
    "        combined = [(ranges[0][0], ranges[0][1])] \n",
    "\n",
    "        for start, end, intensity, sequence in ranges[1:]:\n",
    "            last_start, last_end = combined[-1]\n",
    "            if start <= last_end:\n",
    "                combined[-1] = (last_start, max(last_end, end))\n",
    "            else:\n",
    "                combined.append((start, end))\n",
    "\n",
    "        return combined\n",
    "\n",
    "\n",
    "    def calculate_coverage_percentage(combined_ranges, seq_length):\n",
    "        coverage = sum(end - start for start, end in combined_ranges)\n",
    "        percentage_coverage = (coverage / seq_length) * 100\n",
    "        return percentage_coverage\n",
    "    \n",
    "    def calculate_brightness(color):\n",
    "        # Convert hex to RGB\n",
    "        r, g, b = to_rgb(color)\n",
    "        # 1-255\n",
    "        r, g, b = [int(x * 255) for x in [r, g, b]]\n",
    "        # equation for finding brightness\n",
    "        brightness = (0.299 * r) + (0.587 * g) + (0.114 * b)\n",
    "\n",
    "        return brightness\n",
    "\n",
    "    def generate_html_representation(protein_sequence, coverage_ranges, color=\"red\"):\n",
    "        marked_sequence = \"\"\n",
    "        position = 0\n",
    "\n",
    "        # Calculate brightness and decide text color\n",
    "        brightness = calculate_brightness(color)\n",
    "        text_color = \"black\" if brightness > 128 else \"white\"\n",
    "\n",
    "        for start, end in coverage_ranges:\n",
    "            start = int(start) - 1  \n",
    "            marked_sequence += protein_sequence[position:start]  \n",
    "            marked_sequence += f'<span style=\"background-color:{color}; color:{text_color}\">{protein_sequence[int(start):int(end)]}</span>'  # Covered region\n",
    "            position = int(end)  \n",
    "\n",
    "        marked_sequence += protein_sequence[position:] \n",
    "        return marked_sequence\n",
    "\n",
    "    # Run dependecy functions\n",
    "    prot_colour_mapping={}\n",
    "    #print(sorted(POIs_df[prot_id_col].values.tolist()))\n",
    "    #print(colours)\n",
    "    for majority_prot_list,c in zip(sorted(POIs),colours):\n",
    "        for prot in majority_prot_list.split(\";\"):\n",
    "            prot_colour_mapping[prot]=c\n",
    "\n",
    "    seq_dict = read_fasta(fasta_file)\n",
    "    peptide_df = filter_peptides(peptides_df, exp)\n",
    "    protein_peptide_dict = get_protein_peptides(peptide_df, POIs_df, exp, prot_id_col, pept_id_col)\n",
    "    html_proteins={}\n",
    "\n",
    "    # Plot coverage\n",
    "    for protein, peptides in protein_peptide_dict.items():\n",
    "        if not peptides:\n",
    "            #printmd(f'<span style=\"font-family:Open Sans, verdana, arial, sans-serif;font-size:12;margin-top:0px;margin-bottom:0px;\"No valid peptides for protein: {protein}\"<span>')\n",
    "            continue\n",
    "        combined = combine_ranges(protein_peptide_dict[protein])\n",
    "        seq_length=seq_dict[protein][1]\n",
    "        seq_string=seq_dict[protein][0]\n",
    "        seq_cov=calculate_coverage_percentage(combined, seq_length)\n",
    "\n",
    "        fig = go.Figure()\n",
    "        for s, e in combined:\n",
    "            fig.add_trace(\n",
    "                go.Scatter(x=[s-0.5,s-0.5,e+0.5,e+0.5,s-0.5],y=[0,1,1,0,0],fill=\"toself\",\n",
    "                fillcolor = prot_colour_mapping[protein],\n",
    "                mode='none',line=dict(width=0),showlegend=False,hoverinfo='skip')\n",
    "            )\n",
    "        fig.add_trace(go.Scatter(x=[0.5,0.5,seq_length+0.5,seq_length+0.5,0.5],y=[0,1,1,0,0],\n",
    "                                 line=dict(color='black',width=1.5),\n",
    "                                mode='lines',showlegend=False,hoverinfo='skip')\n",
    "                     )\n",
    "        fig.update_xaxes(range=[0.5, seq_length+1],showgrid=False)\n",
    "        fig.update_yaxes(range=[0, 1],showgrid=False,showticklabels=False,ticklen=0)\n",
    "        fig.update_layout(\n",
    "            title=f'Peptidyl signal coverage for {protein}<br><sup>Sequence coverage: {round(seq_cov,2)}%<sup>', \n",
    "            xaxis_title=\"Amino acid position\",\n",
    "            plot_bgcolor='#EEEAE8',\n",
    "            #legend_title_text='Proteins',\n",
    "            yaxis={'tickformat':\".2\", 'rangemode': 'tozero',\n",
    "                   'ticks': 'outside'},\n",
    "            font=dict(color='black',size=12),\n",
    "            autosize=False,\n",
    "            width=800,\n",
    "            height=200,\n",
    "        )\n",
    "        \n",
    "        config = {\n",
    "          'toImageButtonOptions': {\n",
    "            'format': 'png', # one of png, svg, jpeg, webp\n",
    "            'filename': 'Peptide_coverage_plot',\n",
    "            'scale': 4 # Multiply title/legend/axis/canvas sizes by this factor\n",
    "          }\n",
    "        }\n",
    "        # Show coverage plot\n",
    "        fig.show(config=config)\n",
    "\n",
    "        # Text coverage\n",
    "        #html_proteins[protein] = generate_html_representation(seq_string, combined, prot_colour_mapping[protein])\n",
    "        html_representation = generate_html_representation(\n",
    "            seq_string,\n",
    "            combined, \n",
    "            color=prot_colour_mapping[protein]\n",
    "        )\n",
    "        printmd(\n",
    "        '<div style=\"width:80ch;font-size:14px;font-family:Open Sans, verdana, arial, sans-serif;margin-left:3em;\">'\n",
    "        '<p style=\"word-break: break-all\">'\n",
    "        +html_representation+\n",
    "        '</p>'\n",
    "        '</div>'\n",
    "        )\n",
    "        \n",
    "        # Peptide intensity plot\n",
    "        fig2 = go.Figure()\n",
    "        \n",
    "        intensity_values=[]\n",
    "        \n",
    "        for s, e, inten, pep_sequence in peptides:\n",
    "            if inten > 0:\n",
    "                intensity_values.append(inten)\n",
    "                fig2.add_trace(go.Scatter(x=[s,e],y=[inten,inten],\n",
    "                                         #fill=\"toself\",\n",
    "                        mode='lines',line=dict(color=prot_colour_mapping[protein]),\n",
    "                                         #line=dict(width=100)\n",
    "                                         showlegend=False,\n",
    "                                         hovertemplate =\n",
    "                                        f'Start position: {s}'+\n",
    "                                        f'<br>End position: {e}'+\n",
    "                                        f'<br>Intensity: {inten:e}'+\n",
    "                                        f'<br>Sequence: {pep_sequence}'\n",
    "                                        )\n",
    "                         )\n",
    "                middle_x = (s + e) / 2\n",
    "                fig2.add_trace(\n",
    "                            go.Scatter(x=[middle_x], y=[inten], \n",
    "                            mode='markers', marker=dict(size=0, opacity=0, color=prot_colour_mapping[protein]),\n",
    "                            showlegend=False,\n",
    "                            hovertemplate=\n",
    "                                       f'Start position: {s}'+\n",
    "                                       f'<br>End position: {e}'+\n",
    "                                       f'<br>Intensity: {inten:e}'+\n",
    "                                       f'<br>Sequence: {pep_sequence}'\n",
    "                                      )\n",
    "                            )\n",
    "        \n",
    "        fig2.update_traces(name='')\n",
    "        \n",
    "        max_val = np.max(intensity_values)\n",
    "        min_val = np.min(intensity_values)\n",
    "        # For adjusting plots y-axis\n",
    "        \n",
    "         # Calculate log base 10 of min and max\n",
    "        log_min_val = np.floor(np.log10(min_val)) if min_val > 0 else 0\n",
    "        log_max_val = np.ceil(np.log10(max_val))\n",
    "\n",
    "        # Want to ensure that there's more than one major grid line\n",
    "        if (log_min_val == log_max_val) or (log_max_val - log_min_val <= 2):\n",
    "            log_max_val += 1  # one order of magnitude larger\n",
    "\n",
    "        # get tick values\n",
    "        tick_vals = [10 ** i for i in range(int(log_min_val), int(log_max_val) + 1)]\n",
    "        tick_text = [f\"{x:.0e}\" for x in tick_vals]\n",
    "        \n",
    "        fig2.update_layout(\n",
    "            title=f'Peptide signal intensity for {protein}',\n",
    "            yaxis_title=f'Peptide intensity', \n",
    "            xaxis_title=\"Amino acid position\",\n",
    "            xaxis=dict(\n",
    "            range=[1, seq_length], \n",
    "            ),\n",
    "            plot_bgcolor='#EEEAE8',\n",
    "            font=dict(color='black',size=12),\n",
    "            yaxis=dict(\n",
    "            type=\"log\",\n",
    "            tickvals=tick_vals,\n",
    "            ticktext=tick_text,\n",
    "            range=[np.log10(min(tick_vals)), np.log10(max(tick_vals))]  # Log scale range\n",
    "            )\n",
    "          )\n",
    "        config = {\n",
    "          'toImageButtonOptions': {\n",
    "            'format': 'png',\n",
    "            'filename': 'Peptide_intensity_plot',\n",
    "            'scale': 4\n",
    "          }\n",
    "        }\n",
    "        fig2.show(config=config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "844b4f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of proteins to consider\n",
    "num=10\n",
    "protein_groups = pd.read_excel(excel_path,sheet_name=\"Full_proteinGroups\")\n",
    "peptides_data = pd.read_excel(excel_path,sheet_name=\"Peptides\") \n",
    "    \n",
    "if reattribute_peptides==False:\n",
    "    prot_id_col='Majority protein IDs'\n",
    "    pept_id_col='Proteins'\n",
    "elif reattribute_peptides==\"no\":\n",
    "    prot_id_col='Majority protein IDs'\n",
    "    pept_id_col='Proteins'\n",
    "elif reattribute_peptides==True:\n",
    "    prot_id_col='Reattributed Protein'\n",
    "    pept_id_col='Reattributed Protein'\n",
    "elif reattribute_peptides==\"yes\":\n",
    "    prot_id_col='Reattributed Protein'\n",
    "    pept_id_col='Reattributed Protein'\n",
    "      \n",
    "for experiment in sorted(experiments):\n",
    "    POIs_df=get_POIs_df(experiment,protein_groups,num,POIs)\n",
    "    #updated_POIs=return_new_POIs(POIs,POIs_df)\n",
    "    top_n=protein_df_organization(experiment,protein_groups,num,POIs_df,prot_id_col)\n",
    "    make_report_title(experiment,POIs,identification)\n",
    "    make_subtitle(num,POIs,POIs_df)\n",
    "    topn_proteins_identified(top_n,POIs,POIs_df,num,experiment,colours,prot_id_col)\n",
    "    try:\n",
    "        relevant_peptides_df = filter_peptides(peptides_data, POIs_df, identification,prot_id_col,pept_id_col)\n",
    "        peptide_plots(fasta_file,relevant_peptides_df,POIs,POIs_df,experiment,colours,prot_id_col,pept_id_col)\n",
    "    except KeyError:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e3c2f32",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
